controllers/component_dependency_update_controller.go

```diff
@@ -0,0 +1,484 @@
+		// As tekton results should aggressivly delete when pruning is enabled
+		controllerutil.AddFinalizer(pipelineRun, NudgeFinalizer)
+		err := r.Client.Update(ctx, pipelineRun)
+		if err != nil {
+			log.Error(err, "failed to add finalizer")
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// handleCompletedBuild will perform a 'nudge' updating dependent downstream components
+// This will involve creating a PR updating their references to our images to the newly produced image
+// For now it just produces a log message
+func (r *ComponentDependencyUpdateReconciler) handleCompletedBuild(ctx context.Context, pipelineRun *tektonapi.PipelineRun, updatedComponent *applicationapi.Component) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+	success := pipelineRun.Status.GetCondition(apis.ConditionSucceeded).IsTrue()
+	if !success {
+		log.Info("not performing nudge as pipeline failed")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	// find the image and digest we want to update to
+	image := ""
+	tag := ""
+	digest := ""
+	for _, r := range pipelineRun.Status.Results {
+		if r.Name == ImageDigestParamName {
+			digest = r.Value.StringVal
+		} else if r.Name == ImageUrlParamName {
+			image = r.Value.StringVal
+		}
+	}
+	// Failure to find is a permanent error, so remove the finalizer
+	if image == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageUrlParamName), "no image url result")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	if digest == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageDigestParamName), "no image digest result")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	parts := strings.Split(image, ":")
+	if len(parts) != 2 {
+		log.Error(fmt.Errorf("invalid image name %s", image), "invalid image format")
+		return r.removePipelineFinalizer(ctx, pipelineRun)
+	}
+	tag = parts[1]
+	repo := parts[0]
+
+	components := applicationapi.ComponentList{}
+	err := r.Client.List(ctx, &components, client.InNamespace(pipelineRun.Namespace))
+	if err != nil {
+		log.Error(err, "failed to list components in namespace")
+		return ctrl.Result{}, err
+	}
+
+	retryTime := FailureRetryTime
+	immediateRetry := false
+
+	toUpdate := []applicationapi.Component{}
+
+	for i := range components.Items {
+		comp := components.Items[i]
+		if comp.Spec.Application == updatedComponent.Spec.Application && slices.Contains(updatedComponent.Spec.BuildNudgesRef, comp.Name) {
+			toUpdate = append(toUpdate, comp)
+		}
+	}
+	var nudgeErr error
+	immediateRetry, nudgeErr = r.UpdateFunction(ctx, r.Client, r.Scheme, r.EventRecorder, toUpdate, &BuildResult{BuiltImageRepository: repo, BuiltImageTag: tag, Digest: digest, Component: updatedComponent})
+	if nudgeErr != nil {
+		log.Error(nudgeErr, fmt.Sprintf("component update of as a result of a build of %s failed", updatedComponent.Name), l.Audit, "true")
```

Comment:
`"component update of <something missing here?> as a result of a build of %s failed"`